internal/event/dispatch_service.go

```diff
@@ -100,12 +78,9 @@ func (s *DispatchService) Subscribe(eventName string, subscriptionID int64, hand
 	s.logger.Trace(func() string {
 		return fmt.Sprintf("event.DispatchService.Subscribe: %s, %d, %v", eventName, subscriptionID, handler)
 	})
```

Comment:
Should our Subscribe guarantee the order of events ? 
The doc says `Do not rely on the order of handlers, they may be shuffled.` . Do we mean order of `events` or `handlers` ?
Can we make the behaviour of events more clear on the doc as well ?

Reply:
That's really the order of handlers. If the user registers handlerA then handlerB to an event, then we don't guarantee that handlerA will be dispatched before handlerB.

Events should be received in order, and if the user subscribed using the sync variant, then they should observe it in the same order. I'll add a test for that.

Reply:
I was asking for Subscribe method not SubscribeSync.
Should the order be guaranteed for `Subscribe` ?
And is the order guaranteed for `Subscribe` ?

Reply:
The test is here: https://github.com/hazelcast/hazelcast-go-client/pull/644/commits/2e74a09a6915ca8100eb663f70d081022789e8fa

Reply:
We can't guarantee that easily for Subscribe, since the corresponding handlers are dispatched in a goroutine.

Reply:
Is not guaranteeing the order of `Subscribe` will cause any correctness problems where it is used ?

Reply:
I think we should be using Sync events where the order of events may cause problems.

Reply:
I see that almost all usages are problematic where open/closed added/removed listened via same service. 
https://github.com/hazelcast/hazelcast-go-client/blob/master/internal/cluster/connection_listener_binder.go#L73
https://github.com/hazelcast/hazelcast-go-client/blob/master/internal/cluster/connection_manager.go#L317
https://github.com/hazelcast/hazelcast-go-client/blob/master/internal/cluster/view_listener_service.go#L27

What do you think ? I did a quick scan over the code. I will let you to check the rest.

Reply:
Also note that if you use SubsribeSync in internals, it means that we should be make sure that design is deadlock free.

Reply:
I was thinking how to solve this issue. I can say that making sure the design is deadlock free is hard from experience.
I think if we design the dispatch service in only async way where we guarantee the order, it will be a lot easier to implement the rest. 

What do you think ?

Reply:
Can you explain a bit about what you mean by guaranteeing the order? Should an event handler be dispatched in the order of the "Publish" calls and run to completion before the other event handler starts? Or just starting an event handler (but not waiting for its completion) before the next event handler is adequate?

Sync handlers are required only in a few places and they are just used to publish an event on the user event dispatcher. In another PR, that probably will mean writing to a channel, so that won't block.

Reply:
> Should an event handler be dispatched in the order of the "Publish" calls and run to completion before the other event handler starts? Or just starting an event handler (but not waiting for its completion) before the next event handler is adequate?

The first one. The second one does not add any value. In the eyes of the caller of the service, the two are still concurrent. 


Lets say that I do two Publishes where there is an order between them, lets say that from the single thread to make it simple. It could be a lot more complicated(like two events coming from remote where the order is guaranteed by the rest of the system). 

service.Publish(MemberConnected(address))
service.Publish(MemberDisconnected(address))

//lets say these two are subscribed to dispatchService 
func connectedHandler(event) {
    //TAG1
    //.... Users code
}

func disconnectedHandler(event) {
    //TAG2
    //.... Users code
}

Lets say that we are doing the second one and `not waiting for its completion`
1. The connected handler is called, we made sure it is started. That means `disconnectedHandler` can be called via dispathService . 
2. At the beginning of `connectedHandler` , at TAG1, we hit a garbage collection and go decided to park our thread. 
3. dispathService calls disconnectedHandler. 
4. disconnectedHandler is finished. 
5. Go decided that now it is ok to let `connectedHandler` to continue from TAG1. 

As you can see `disconnectedHandler` worked before `connectedHandler`

Reply:
If we wait the end of a handler to completion before launching another handler, then it's not asynchronous. I don't think there's a solution that we can apply to all cases. But for this specific issue, how about making sure that the corresponding connection in disconnected handler is connected before going with disconnected logic? Maybe having a separate event which is fired only after `ConnectionOpened` is completed?

Reply:
> If we wait the end of a handler to completion before launching another handler, then it's not asynchronous
We can still make it async in the sense that caller of publish will not be blocked. I think that is enough right ?

>  how about making sure that the corresponding connection in disconnected handler is connected before going with disconnected logic? Maybe having a separate event which is fired only after ConnectionOpened is completed?

We can always have more complex solutions. But I believe there is a much simpler solution.
We still focus on a simple dispatch service where 
1- It is async for the caller of Publish. Meaning Publish will not be blocked. 
2- Events fired from the same thread will be handled in the same order. One will finish, then other will start.  (it is enough to make sure this per EventName)
3 - If we block an event, it should not block all events only ones that are related to same EventName. 
4 - [OPTIONALLY] a close after publish in the same thread waits for published item to be handled(finished) . 

I think implementing a dispatch service guaranteeing the at least first three should be simple. I think the forth item will come free after that but lets wait and see. 
And it looks like what all we want. 
What do you think ?

Reply:
> Events fired from the same thread will be handled in the same order

Currently events may be published from different goroutines, E.g,. ConnectionOpened is published from connection manager, ConnectionClosed is published from socket read/write loops and the connection manager.

I think the problem is not with the goroutine that publishes the events but with the subscribers. E.g., we want to make sure that ConnectionClosed handler for a connection starts iff the corresponding ConnectionOpened handler ran to completion.  How about keeping a set of connections for which the ConnectionOpened handler ran to completion? Then ConnectionClosed handlers would block until the corresponding connection is in that set.

Reply:
I does not matter whether ConnectionOpened and ConnectionClosed is published from different threads. What matters is that is there a happens-before relation between them or not ? I believe there is. 

Again. We can solve a simple problem once in the dispatch service, or we can solve this every place again and again that uses this dispatch service. I prefer solving it once. 

As I said here https://github.com/hazelcast/hazelcast-go-client/pull/644#discussion_r713845921
Do you think we can find this simple solution , or am I asking something impossible ?
 Or would it be ok if I try to send a pr for it ?

Reply:
We both agree that ConnectionClosed handler should be executed after ConnectionOpened. Of course, please send a PR if there is an easier way of accomplishing that than I described.

Reply:
Here it is 
https://github.com/hazelcast/hazelcast-go-client/pull/652